src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (let j = 0; j < action.keysToRemove.length; j++) {
+        const key = action.keysToRemove[j];
+
+        delete currentState[key];
+      }
+    }
```

Comment:
Consider using a switch statement instead of multiple if-else blocks since you have a limited and known set of action types. This will make the code cleaner and more maintainable.